src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -955,6 +936,243 @@ void C2_MacroAssembler::fast_unlock(Register objReg, Register boxReg, Register t
   bind(NO_COUNT);
 }
 
+void C2_MacroAssembler::fast_lock_lightweight(Register obj, Register box, Register rax_reg,
+                                              Register t, Register thread) {
+  assert(LockingMode == LM_LIGHTWEIGHT, "must be");
+  assert(rax_reg == rax, "Used for CAS");
+  assert_different_registers(obj, box, rax_reg, t, thread);
+
+  // Handle inflated monitor.
+  Label inflated;
+  // Finish fast lock successfully. ZF value is irrelevant.
+  Label locked;
+  // Finish fast lock unsuccessfully. MUST jump with ZF == 0
+  Label slow_path;
+
+  if (DiagnoseSyncOnValueBasedClasses != 0) {
+    load_klass(rax_reg, obj, t);
+    movl(rax_reg, Address(rax_reg, Klass::access_flags_offset()));
+    testl(rax_reg, JVM_ACC_IS_VALUE_BASED_CLASS);
+    jcc(Assembler::notZero, slow_path);
+  }
+
+  const Register mark = t;
+
+  { // Lightweight Lock
+
+    Label push;
+
+    const Register top = box;
+
+    // Prefetch the mark.
+    movptr(mark, Address(obj, oopDesc::mark_offset_in_bytes()));
+
+    // Check if lock-stack is full.
+    cmpl(Address(thread, JavaThread::lock_stack_top_offset()), LockStack::end_offset() - 1);
+    jcc(Assembler::greater, slow_path);
+
+    // Check if recursive.
+    movl(top, Address(thread, JavaThread::lock_stack_top_offset()));
+    cmpptr(obj, Address(thread, top, Address::times_1, -oopSize));
+    jccb(Assembler::equal, push);
+
+    // Check for monitor (0b10).
```

Comment:
I believe you are correct. In fast_lock it might be better to check for monitor first. 

Have to run some more benchmarks. But running some (single threaded un-contended) micros show that moving monitor check in lock has modest improvements on inflated locking and very minor, if any regressions on lightweight locking. 

The scaling was mostly seen when moving the mark word load in the unlock case. In fast unlock the lock stack must be checked first for correctness if we wish to elide the owner field anonymous owner check.